### PR TITLE
Add Logging component

### PR DESCRIPTION
**Pull request by Half-Shot**
This one is based off gitter's logger. It optionally depends on winston but will use the console if winston is not available.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -94,6 +94,41 @@ The component which orchestrates other components: a "glue" component. Provides
 a way to start the bridge. This is the component most examples use. Has
 dependencies on most of the components listed above.
 
+## `Logging`
+This component exposes access to the bridges log reporter. To use, you should
+install the optional packages `winston`, `winston-daily-rotate-file` and
+`chalk` to get nice formatted log lines, otherwise it will default to the JS
```
Comment (turt2live) [inline README.md]: we might want to recommend versions here, like `winston@3` or whatever

File: README.md [hunk 2]
```diff
@@ -94,6 +94,41 @@ The component which orchestrates other components: a "glue" component. Provides
...
+console. To use the component, use `Logging.Configure(configObject)` to setup
+the logger, which takes the following options:
+```
+    # A level to set the console reporter to.
+    console: "error|warn|info|debug|off"
+
+    # Format to append to log files.
+    fileDatePattern: "YYYY-MM-DD",
```
Comment (turt2live) [inline README.md]: the others don't have a comma, and they probably should. Can we get away with putting the surrounding curly braces around the whole snippet and syntax highlight it for javascript? This is purely aesthetic and nothing else

File: README.md [hunk 3]
```diff
@@ -94,6 +94,41 @@ The component which orchestrates other components: a "glue" component. Provides
...
+
+    # The maximum number of files per level before old files get cleaned
+    up. Use 0 to disable.
+    maxFiles: 5
+```
+
+You **MUST** configure the logger before anything will be emitted to the console.
+
+You can then use `Logging.Get(ModuleName)` to start logging to the reporter,
+using the `error, warn, info, debug` functions. Arguments to these functions will
```
Comment (turt2live) [inline README.md]: nitpick: `error`, `warn`, etc (don't include the comma in the code tag)

File: lib/bridge.js [hunk 4]
```diff
@@ -18,6 +18,7 @@ const Promise = require("bluebird");
 const Datastore = require("nedb");
 const util = require("util");
 const MembershipCache = require("./components/membership-cache");
+const log = require("./components/logging").Get("Bridge");
```
Comment (turt2live) [inline lib/bridge.js]: general principle: the log name should match the file name including casing (`bridge`, `cli`, etc)

File: README.md [hunk 5]
```diff
@@ -94,6 +94,41 @@ The component which orchestrates other components: a "glue" component. Provides
...
+    # A level to set the console reporter to.
+    console: "error|warn|info|debug|off"
+
+    # Format to append to log files.
+    fileDatePattern: "YYYY-MM-DD",
+
+    # Format of the timestamp in log files.
+    timestampFormat: "MMM-D HH:mm:ss.SSS"
+
+    # Log files to emit to, keyed of the minimum level they report.
```
Comment (turt2live) [inline README.md]: Should also say that not including the option (or having it falsey) disables files

File: README.md [hunk 6]
```diff
@@ -94,6 +94,41 @@ The component which orchestrates other components: a "glue" component. Provides
...
+    files: {
+        "abc.log" => "error|warn|info|debug|off"
```
Comment (turt2live) [inline README.md]: it's actually `"abc.log": "error|warn|etc"`?

File: README.md (hunk 1 above)
Comment (Half-Shot) [reply README.md]: I kinda hoped it would be implied by the package.json, as I think you can ask npm to install optionals.

Comment (turt2live) [reply README.md]: I tried to find information on npm installing deps that nested in a project, but couldn't find anything. Duplicating the versions (even just the major versions) is fine here, I think.

File: README.md (hunk 6 above)
Comment (Half-Shot) [reply README.md]: Or "/tmp/abc.log", it can be relative.

Comment (turt2live) [reply README.md]: I mean more that the arrow is not the right syntax, is it?

**Pull request merged**